src/Domain/CoordinateMaps/TimeDependent/Shape.cpp

```diff
@@ -0,0 +1,338 @@
+  const auto& x = cartesian[0];
+  const auto& y = cartesian[1];
+  const auto& z = cartesian[2];
+  return {atan2(hypot(x, y), z), atan2(y, x)};
+}
+
+template <typename TransitionFunc>
+Shape<TransitionFunc>::Shape(const std::array<double, 3>& center, size_t l_max,
+                             size_t m_max, TransitionFunc transition_func,
+                             std::string function_of_time_name) noexcept
+    : f_of_t_name_(std::move(function_of_time_name)),
+      center_(center),
+      l_max_(l_max),
+      m_max_(m_max),
+      ylm_(l_max, m_max),
+      transition_func_(std::move(transition_func)) {
+  ASSERT(l_max >= 2, "The shape map requires l_max >= 2 but l_max = " << l_max);
+  ASSERT(m_max >= 2, "The shape map requires m_max >= 2 but m_max = " << m_max);
+}
+
+template <typename TransitionFunc>
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> Shape<TransitionFunc>::operator()(
+    const std::array<T, 3>& source_coords, double time,
+    const FunctionsOfTimeMap& functions_of_time) const noexcept {
+  ASSERT(functions_of_time.find(f_of_t_name_) != functions_of_time.end(),
+         "Could not find function of time: '"
+             << f_of_t_name_ << "' in functions of time. Known functions are "
+             << keys_of(functions_of_time));
+
+  const auto centered_coords = center_coordinates_(source_coords);
+  auto theta_phis = cartesian_to_spherical(centered_coords);
+  const auto interpolation_info = ylm_.set_up_interpolation_info(theta_phis);
+  const DataVector coefs = functions_of_time.at(f_of_t_name_)->func(time)[0];
+  check_coefficients_(coefs);
+  // re-use allocation
+  auto& distorted_radii = get<0>(theta_phis);
+  // evaluate the spherical harmonic expansion at the angles of `source_coords`
+  ylm_.interpolate_from_coefs(make_not_null(&distorted_radii), coefs,
+                              interpolation_info);
+  return center_ +
+         centered_coords *
+             (1. - distorted_radii * transition_func_(centered_coords));
+}
+
+template <typename TransitionFunc>
+std::optional<std::array<double, 3>> Shape<TransitionFunc>::inverse(
+    const std::array<double, 3>& target_coords, double time,
+    const FunctionsOfTimeMap& functions_of_time) const noexcept {
+  ASSERT(functions_of_time.find(f_of_t_name_) != functions_of_time.end(),
+         "Could not find function of time: '"
+             << f_of_t_name_ << "' in functions of time. Known functions are "
+             << keys_of(functions_of_time));
+
+  const auto centered_coords = center_coordinates_(target_coords);
+  const auto theta_phis = cartesian_to_spherical(centered_coords);
+  const DataVector coefs = functions_of_time.at(f_of_t_name_)->func(time)[0];
+  check_coefficients_(coefs);
+  const auto distorted_radii = ylm_.interpolate_from_coefs(coefs, theta_phis);
+  return center_ +
+         centered_coords *
+             (1. + distorted_radii * transition_func_(centered_coords));
+}
```

Comment:
Here it looks like you are implementing the equation
`\xi^i = x^i + (x^i - x_c^i) f(\tilde{r}, \theta, \phi) \sum_{lm} \lambda_{lm}(t)Y_{lm}(\theta, \phi)`
instead of what is in the dox in the hpp file, which is
`\xi^i = x^i + (x^i - x_c^i) f(r, \theta, \phi) \sum_{lm} \lambda_{lm}(t)Y_{lm}(\theta, \phi)`.

The difference between the two expressions is the radius that you pass into `f`. In the code you are passing the radius that I call `\tilde{r}`, which is the magnitude of `x^i-x_c^i`, but in the dox you pass in `r`, which is the magnitude of `xi^i-x_c^i`.  The two different radii are related by
`\tilde{r} = r (1 - f(r,\theta,\phi) \sum_{lm} \lambda_{lm}(t)Y_{lm}(\theta, \phi) )`.

I don't think either expression (the one in the dox, or the one in the code) is correct for the inverse.  I think one correct expression is
`\xi^i = x^i + (xi^i - x_c^i) f(r, \theta, \phi) \sum_{lm} \lambda_{lm}(t)Y_{lm}(\theta, \phi)`.  But evaluating this is a problem, because you don't know the `r` to pass into `f` because you don't yet know `\xi^i`.

By the way, my expression and the expression you evaluate in the code happen to be equal only for the case when `f(r,\theta,\phi) = 1/r`.  Which is exactly the `f(r,\theta\phi)` in SphereTransition.  I think this is why your test passes; it should fail if any other TransitionFunction is chosen.

The procedure that I came up with in SpEC to evaluate the inverse is the following:  The TransitionFunction has an additional member function `r_over_tilde_r`, which takes as arguments `x^i-x_c^i` at a single point and `\lambda = \sum_{lm} \lambda_{lm}(t)Y_{lm}(\theta, \phi)` at that same point.  The function then computes `\tilde{r}` trivially from the first argument it is passed, and then computes `r` using it's inside knowledge of what `f` is, and returns `r/\tilde{r}`.  Then the inverse function evaluates `xi^i = x_c^i + (x^i-x_c^i)*(r/\tilde{r})`, using the fact that the angles are preserved by the map.